File: src/org/cappuccino_project/ide/intellij/plugin/psi/interfaces/ObjJFunctionDeclarationElement.java
package org.cappuccino_project.ide.intellij.plugin.psi.interfaces;

import org.cappuccino_project.ide.intellij.plugin.psi.ObjJBlock;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJFormalParameterArg;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJLastFormalParameterArg;
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJFunctionDeclarationElementStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public interface ObjJFunctionDeclarationElement<StubT extends ObjJFunctionDeclarationElementStub<? extends ObjJFunctionDeclarationElement>> extends ObjJHasFunctionName, ObjJStubBasedElement<StubT>, ObjJCompositeElement, ObjJResolveableElement<StubT> {
    @NotNull
    List<String> getParamNames();
    @Nullable
    String getReturnType();

    @NotNull
    List<ObjJFormalParameterArg> getFormalParameterArgList();

    @Nullable
    ObjJLastFormalParameterArg getLastFormalParameterArg();

    @Nullable
    ObjJBlock getBlock();
}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/interfaces/ObjJFunctionDeclarationElement.kt
package org.cappuccino_project.ide.intellij.plugin.psi.interfaces

import com.intellij.psi.StubBasedPsiElement
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJBlock
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJFormalParameterArg
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJLastFormalParameterArg
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJFunctionDeclarationElementStub

interface ObjJFunctionDeclarationElement<StubT: ObjJFunctionDeclarationElementStub<*>> : ObjJHasFunctionName, ObjJStubBasedElement<StubT>, ObjJCompositeElement, ObjJResolveableElement<StubT> {
    val paramNames: List<String>
    val returnType: String?

    val formalParameterArgList: List<ObjJFormalParameterArg>

    val lastFormalParameterArg: ObjJLastFormalParameterArg?

    val block: ObjJBlock?
}
